drivers/regulator/regulator_pca9422.c

```diff
@@ -0,0 +1,1234 @@
+					       uint8_t reg_addr, uint8_t mask, uint8_t value)
+{
+	int ret;
+
+	/* have to unlock first or else register read returns 0x0 */
+	ret = regulator_pca9422_unlock(spec);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* do the update while unlocked */
+	ret = i2c_reg_update_byte(spec->bus, spec->addr, reg_addr, mask, value);
+
+	/* always unlock and return result of the update */
+	regulator_pca9422_lock(spec);
+
+	return ret;
+}
+
+static int regulator_pca9422_read_locked_reg(const struct i2c_dt_spec *spec,
+					     uint8_t reg_addr, uint8_t *value)
+{
+	int ret;
+
+	/* have to unlock first or else register read returns 0x0 */
+	ret = regulator_pca9422_unlock(spec);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* do the read while unlocked */
+	ret = i2c_reg_read_byte(spec->bus, spec->addr, reg_addr, value);
+
+	/* always unlock and return result of the update */
+	regulator_pca9422_lock(spec);
+
+	return ret;
+}
+
+static int regulator_pca9422_set_voltage(const struct device *dev,
+					 int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_pca9422_config *config = dev->config;
+	const struct regulator_pca9422_common_config *cconfig = config->parent->config;
+	struct regulator_pca9422_common_data *cdata = config->parent->data;
+	uint16_t idx;
+	int ret;
+	uint8_t mode = cdata->dvs_state;
+
+	/* DPSTANDBY mode uses the same voltage as STANDBY */
+	if (mode == PCA9422_DPSTANDBY_MODE) {
+		mode = PCA9422_STANDBY_MODE;
+	}
+
+	ret = linear_range_group_get_win_index(config->desc->ranges,
+					       config->desc->num_ranges, min_uv,
+					       max_uv, &idx);
+	if (ret == -EINVAL) {
+		return ret;
+	}
+
+	return regulator_pca9422_update_locked_reg(&cconfig->i2c,
+						   config->desc->vsel_mode[mode].vsel_reg,
+						   config->desc->vsel_mode[mode].vsel_mask,
+						   (uint8_t)idx);
+}
+
+static int regulator_pca9422_get_voltage(const struct device *dev,
+					 int32_t *volt_uv)
+{
+	const struct regulator_pca9422_config *config = dev->config;
+	const struct regulator_pca9422_common_config *cconfig = config->parent->config;
+	struct regulator_pca9422_common_data *cdata = config->parent->data;
+	int ret;
+	uint8_t raw_reg;
+	uint8_t mode = cdata->dvs_state;
+
+	/* DPSTANDBY mode uses the same voltage as STANDBY */
+	if (mode == PCA9422_DPSTANDBY_MODE) {
+		mode = PCA9422_STANDBY_MODE;
+	}
+
+	ret = regulator_pca9422_read_locked_reg(&cconfig->i2c,
+						config->desc->vsel_mode[mode].vsel_reg,
+						&raw_reg);
+	if (ret < 0) {
+		return ret;
+	}
+
+	raw_reg = (raw_reg & config->desc->vsel_mode[mode].vsel_mask);
+
+	return linear_range_group_get_value(config->desc->ranges,
+					    config->desc->num_ranges, raw_reg,
+					    volt_uv);
+}
+
+static int regulator_pca9422_get_current_limit(const struct device *dev,
+					       int32_t *curr_ua)
+{
+	const struct regulator_pca9422_config *config = dev->config;
+	const struct regulator_pca9422_common_config *cconfig = config->parent->config;
+
+	if (cconfig->vin_ilim_ua == 0U) {
```

Comment:
vin_ilim_ua comes from an enum, so can't be zero

Reply:
Thanks for catching this.